baseapp/grpcserver.go

```diff
@@ -38,31 +40,51 @@ func (app *BaseApp) RegisterGRPCServer(server gogogrpc.Server) {
 			if err != nil {
 				return nil, sdkerrors.Wrapf(
 					sdkerrors.ErrInvalidRequest,
-					"Baseapp.RegisterGRPCServer: invalid height header %q: %v", grpctypes.GRPCBlockHeightHeader, err)
+					"baseapp.RegisterGRPCServer: invalid height header %q: %v", grpctypes.GRPCBlockHeightHeader, err)
 			}
 			if err := checkNegativeHeight(height); err != nil {
 				return nil, err
 			}
 		}
 
-		// Create the sdk.Context. Passing false as 2nd arg, as we can't
-		// actually support proofs with gRPC right now.
-		sdkCtx, err := app.createQueryContext(height, false)
+		reqProto, ok := req.(proto.Message)
+		if !ok {
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "expected %T, got %T", (proto.Message)(nil), req)
+		}
+		reqBz, err := proto.Marshal(reqProto)
 		if err != nil {
-			return nil, err
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrInvalidRequest, "cannot proto marshal: %v", err)
 		}
 
-		// Attach the sdk.Context into the gRPC's context.Context.
-		grpcCtx = context.WithValue(grpcCtx, sdk.SdkContextKey, sdkCtx)
+		abciReq := abci.RequestQuery{
+			Path:   info.FullMethod,
+			Data:   reqBz,
+			Height: height,
+			Prove:  false, // Enable here: https://github.com/cosmos/cosmos-sdk/issues/7036
+		}
+		abciRes, err := clientCtx.QueryABCI(abciReq)
+		if err != nil {
+			return nil, err
+		}
 
-		// Add relevant gRPC headers
-		if height == 0 {
-			height = sdkCtx.BlockHeight() // If height was not set in the request, set it to the latest
+		err = protoCodec.Unmarshal(abciRes.Value, resp)
```

Comment:
I have an error here:
```
rpc error: code = Internal desc = interface conversion: interface is nil, not protoiface.MessageV1
```

and indeed `resp` is of type `nil`. Why isn't it of type `SomethingResponse`?

Reply:
Yeah `resp` would be `nil` because it hasn't been passed in. Instead we need to lookup the `resp` type based on the method name. This can be done by first iterating over all the methods in initialization and building a map of response types.